File: server/src/models/RoomMember.ts
import { Model, DataTypes } from 'sequelize';
import sequelize from '../db/connection';
import Room from './Room';
import User from './User';

class RoomMember extends Model {
  declare id: number;
  declare roomId: number;
  declare userId: number;
  declare isAdmin: boolean;
  declare canGrantAdmin: boolean;
  declare online: boolean;
  declare User?: User;
}

RoomMember.init({
  id: {
    type: DataTypes.INTEGER,
    primaryKey: true,
    autoIncrement: true
  },
  roomId: {
    type: DataTypes.INTEGER,
    allowNull: false,
    references: {
      model: Room,
      key: 'id'
    }
  },
  userId: {
    type: DataTypes.INTEGER,
    allowNull: false,
    references: {
      model: User,
      key: 'id'
    }
  },
  isAdmin: {
    type: DataTypes.BOOLEAN,
    defaultValue: false
  },
  canGrantAdmin: {
    type: DataTypes.BOOLEAN,
    defaultValue: false
  },
  online: {
    type: DataTypes.BOOLEAN,
    defaultValue: false
  }
}, {
  sequelize,
  modelName: 'RoomMember',
  tableName: 'room_members',
  timestamps: false
});

RoomMember.belongsTo(User, { foreignKey: 'userId' });

export default RoomMember; 

File: server/internal/models/room_member.go
package models

import (
	"gorm.io/gorm"
)

// RoomMember represents the relationship between users and rooms
type RoomMember struct {
	ID             uint           `gorm:"primaryKey;autoIncrement" json:"id"`
	RoomID         uint           `gorm:"not null;index" json:"roomId"`
	UserID         uint           `gorm:"not null;index" json:"userId"`
	IsAdmin        bool           `gorm:"default:false" json:"isAdmin"`
	CanGrantAdmin  bool           `gorm:"default:false" json:"canGrantAdmin"`
	Online         bool           `gorm:"default:false" json:"online"`
	DeletedAt      gorm.DeletedAt `gorm:"index" json:"-"`

	// Associations
	User *User `gorm:"foreignKey:UserID" json:"user,omitempty"`
	Room *Room `gorm:"foreignKey:RoomID" json:"room,omitempty"`
}

// TableName specifies the table name for RoomMember model
func (RoomMember) TableName() string {
	return "room_members"
}
